streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java

```diff
@@ -115,8 +162,189 @@ public void shutdown() throws IOException {
+        }
+
+        @Override
+        public void run() {
+            this.myStream.start();
+
+        }
+
+        public void close() {
+            this.myStream.close();
+        }
+
+        public final KafkaStreams getStream() {
+            return myStream;
+        }
+    }
+
+    private void verifyAllKVKeys(StreamRunnable[] streamRunnables, KafkaStreams streams,
+                                 Set<String> keys, Set<String> stateStores) throws Exception {
+        for (String storeName : stateStores) {
+            for (String key : keys) {
+
+                // first query where the key is located
+                StreamsMetadata metadata = null;
+                ReadOnlyKeyValueStore<String, Long> store = null;
+                Long value = null;
+                int retries = 0;
+                while (retries < QSRETRIES && (metadata == null || store == null || value == null)) {
+                    try {
+                        metadata = streams.metadataForKey(storeName, key, new StringSerializer());
+                        if (metadata == null) {
+                            retries++;
+                            Thread.sleep(QSBACKOFF);
+                            continue;
+                        }
+                        int index = metadata.hostInfo().port();
+
+                        // next make sure the value is queryable
+                        KafkaStreams streamsWithKey = streamRunnables[index].getStream();
+                        store = streamsWithKey.store(storeName, QueryableStoreTypes.<String, Long>keyValueStore());
+                        if (store == null) {
+                            retries++;
+                            Thread.sleep(QSBACKOFF);
+                            continue;
+                        }
+                        value = store.get(key);
+                        if (value == null) {
+                            retries++;
+                            Thread.sleep(QSBACKOFF);
+                            continue;
+                        }
+                        retries++;
+                    } catch (Exception e) {
+                        retries++;
+                        Thread.sleep(QSBACKOFF);
+                    }
+                }
+                assertThat(metadata, notNullValue());
+                assertThat(store, notNullValue());
+                assertThat(value, notNullValue());
+            }
+        }
+    }
+
+    @Test
+    public void queryOnRebalance() throws Exception {
+        int numThreads = 2;
+        StreamRunnable[] streamRunnables = new StreamRunnable[numThreads];
+        Thread[] streamThreads = new Thread[numThreads];
+        final int numIterations = 500000;
+
+        // create concurrent producer
+        ProducerRunnable producerRunnable = new ProducerRunnable(STREAM_THREE, inputValues, numIterations);
+        Thread producerThread = new Thread(producerRunnable);
```

Comment:
You probably should make sure the `producerThread` is stopped at the end of the test. Perhaps make it a field and in `shutdown` `interrupt` and `join` the thread if it is non-null?